### DISCUSSION TITLE
extern keyword: Is there another purpose than only [DllImport]?

### DISCUSSION THREAD

**Discussion by msedi** [dotnet/runtime#59882] [Q&A] [answered]
I'm currently working with source generators to create all sorts of math operation. For that reason I'm using partial static methods and create the body in the source generator.

I was just curious if I can use the extern keyword for creating something similar like a partial method and assign the funtion pointer there. Or I could use function pointers. 

Even if this is not the suggested way, a question came up of someone else other then the runtime can do something with the extern method other than pinvoke?

**Comment by gfoidl** [comment, marked as answer]
Hopefully [extern (C# Reference)](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/extern) (and the link to the language specification there) can answer your question.

> The `extern` modifier is used to declare a method that is implemented externally. 

Where "externally" is to be understood as outside of the C# code. So a source generator doesn't fall into this category.

**Reply by msedi** [reply]
@gfoidl: Thanks for pointing this out, although I already knew it ;-) I was just curious if I as a regular developer have another thing to do other than the DllImport. What does "defined externally" actually mean? Currently I can only see that DllImport does the assignment but no one else can do anything with the extern method, right?

I was just curious, if I could have used an extern method, other than using unmanaged function pointers? But if no one else can assign something, it seems that it's really only useful for DllImport, or did I muss something?

**Reply by gfoidl** [reply]
> What does "defined externally" actually mean?

That it's not defined in a managed language. You can call methods written in F#, VB.NET from C#-code, as they're all compiled to IL code which the manged runtime, and especially the JIT, can understand.
`extern` methods don't fall into this category, like methods from a native DLL (e.g. written in C). 
So in order for C#-code to call these methods, you have to declare the shape of that method, and that's done by providing the signature for that method with the modifier `extern`, and the DllImport-attribute to specify where this method originates.

> Currently I can only see that DllImport does the assignment but no one else can do anything with the extern method, right?

The runtime could provide other mechanism that provide the external method, but typically -- and for the regular developer -- it's solely DllImport.

> or did I muss something?

I believe you don't miss anything.

**Reply by msedi** [reply]
@gfoidl: Thanks for pointing this out. This helps a lot.
